File: backend/app/services/memory_service.py
import uuid
from datetime import datetime
from typing import List, Optional, Tuple
from sqlalchemy.orm import Session
from app.models.conversation import Conversation
from app.models.message import Message


def get_or_create_conversation(
    user_id: int,
    conversation_id: Optional[str] = None,
    course_id: Optional[int] = None,
    db: Session = None,
) -> Conversation:
    if conversation_id:
        conv = (
            db.query(Conversation)
            .filter(Conversation.id == conversation_id, Conversation.user_id == user_id)
            .first()
        )
        if conv:
            if course_id and not conv.course_id:
                conv.course_id = course_id
                db.commit()
            return conv

    if course_id:
        conv = (
            db.query(Conversation)
            .filter(Conversation.user_id == user_id, Conversation.course_id == course_id)
            .order_by(Conversation.created_at.desc())
            .first()
        )
        if conv:
            return conv

    # Create new
    conv = Conversation(
        id=str(uuid.uuid4()),
        user_id=user_id,
        course_id=course_id,
        created_at=datetime.utcnow(),
    )
    db.add(conv)
    db.commit()
    db.refresh(conv)
    return conv


def get_history(
    conversation_id: str,
    user_id: int,
    db: Session,
    max_messages: int = 10,
) -> List[Tuple[str, str]]:
    conv = (
        db.query(Conversation)
        .filter(Conversation.id == conversation_id, Conversation.user_id == user_id)
        .first()
    )
    if not conv:
        return []

    messages = (
        db.query(Message)
        .filter(Message.conversation_id == conversation_id)
        .order_by(Message.created_at.desc())
        .limit(max_messages)
        .all()
    )

    # Return chronological
    return [(m.role, m.text) for m in reversed(messages)]


def add_message(
    conversation_id: str,
    user_id: int,
    role: str,
    text: str,
    db: Session,
) -> Message:
    conv = (
        db.query(Conversation)
        .filter(Conversation.id == conversation_id, Conversation.user_id == user_id)
        .first()
    )
    if not conv:
        conv = Conversation(
            id=conversation_id,
            user_id=user_id,
            created_at=datetime.utcnow(),
        )
        db.add(conv)
        db.commit()

    msg = Message(
        conversation_id=conversation_id,
        role=role,
        text=text,
        created_at=datetime.utcnow(),
    )
    db.add(msg)
    db.commit()
    db.refresh(msg)
    return msg


def get_course_conversation(user_id: int, course_id: int, db: Session) -> Optional[Conversation]:
    return (
        db.query(Conversation)
        .filter(Conversation.user_id == user_id, Conversation.course_id == course_id)
        .order_by(Conversation.created_at.desc())
        .first()
    )


def delete_course_conversation(user_id: int, course_id: int, db: Session) -> bool:
    convs = (
        db.query(Conversation)
        .filter(Conversation.user_id == user_id, Conversation.course_id == course_id)
        .all()
    )
    for c in convs:
        db.delete(c)
    db.commit()
    return True

File: backend/Services/MemoryService.cs
using Microsoft.EntityFrameworkCore;
using VideoIntelligencePlatform.Backend.Data;
using VideoIntelligencePlatform.Backend.Models;

namespace VideoIntelligencePlatform.Backend.Services;

public class MemoryService : IMemoryService
{
    private readonly AppDbContext _db;

    public MemoryService(AppDbContext db)
    {
        _db = db;
    }

    public async Task<Conversation> GetOrCreateConversationAsync(int userId, string? conversationId, int? courseId = null)
    {
        if (!string.IsNullOrWhiteSpace(conversationId))
        {
            var conversation = await _db.Conversations
                .FirstOrDefaultAsync(c => c.Id == conversationId && c.UserId == userId);

            if (conversation != null)
            {
                if (courseId.HasValue && !conversation.CourseId.HasValue)
                {
                    conversation.CourseId = courseId.Value;
                    await _db.SaveChangesAsync();
                }
                return conversation;
            }
        }

        if (courseId.HasValue)
        {
            var existingCourseConv = await _db.Conversations
                .Where(c => c.UserId == userId && c.CourseId == courseId.Value)
                .OrderByDescending(c => c.CreatedAt)
                .FirstOrDefaultAsync();

            if (existingCourseConv != null)
            {
                return existingCourseConv;
            }
        }

        var newConv = new Conversation
        {
            Id = Guid.NewGuid().ToString(),
            UserId = userId,
            CourseId = courseId,
            CreatedAt = DateTime.UtcNow
        };

        _db.Conversations.Add(newConv);
        await _db.SaveChangesAsync();
        return newConv;
    }

    public async Task<Conversation?> GetCourseConversationAsync(int userId, int courseId)
    {
        return await _db.Conversations
            .Include(c => c.Messages)
            .Where(c => c.UserId == userId && c.CourseId == courseId)
            .OrderByDescending(c => c.CreatedAt)
            .FirstOrDefaultAsync();
    }

    public async Task<bool> ClearCourseConversationAsync(int userId, int courseId)
    {
        var convs = await _db.Conversations
            .Include(c => c.Messages)
            .Where(c => c.UserId == userId && c.CourseId == courseId)
            .ToListAsync();

        if (!convs.Any()) return false;

        foreach (var conv in convs)
        {
            _db.Messages.RemoveRange(conv.Messages);
            _db.Conversations.Remove(conv);
        }

        await _db.SaveChangesAsync();
        return true;
    }

    public async Task<List<ChatHistoryMessage>> GetHistoryAsync(string conversationId, int userId, int maxHistory = 20)
    {
        var messages = await _db.Messages
            .Include(m => m.Conversation)
            .Where(m => m.ConversationId == conversationId && m.Conversation!.UserId == userId)
            .OrderByDescending(m => m.CreatedAt)
            .ThenByDescending(m => m.Id)
            .Take(maxHistory)
            .ToListAsync();

        return messages
            .OrderBy(m => m.CreatedAt)
            .Select(m => new ChatHistoryMessage { Role = m.Role, Text = m.Text })
            .ToList();
    }

    public async Task AddMessageAsync(string conversationId, int userId, string role, string text)
    {
        var conversation = await _db.Conversations
            .FirstOrDefaultAsync(c => c.Id == conversationId && c.UserId == userId);

        if (conversation == null) return;

        var message = new Message
        {
            ConversationId = conversationId,
            Role = role,
            Text = text,
            CreatedAt = DateTime.UtcNow
        };

        _db.Messages.Add(message);
        await _db.SaveChangesAsync();
    }
}
